File: dotnet/src/SmooAI.Observability/Auth/TokenProvider.cs
using System.Net.Http;
using System.Text.Json;

namespace SmooAI.Observability.Auth;

/// <summary>
/// Options for <see cref="TokenProvider"/>.
/// </summary>
public sealed class TokenProviderOptions
{
    /// <summary>OAuth issuer base URL (no trailing slash required). E.g. <c>https://auth.smoo.ai</c>.</summary>
    public string AuthUrl { get; set; } = string.Empty;

    /// <summary>OAuth client id.</summary>
    public string ClientId { get; set; } = string.Empty;

    /// <summary>OAuth client secret (the <c>sk_*</c> minted by Smoo's M2M flow).</summary>
    public string ClientSecret { get; set; } = string.Empty;

    /// <summary>Seconds before expiry to proactively refresh. Defaults to 60s.</summary>
    public int RefreshWindowSec { get; set; } = 60;
}

/// <summary>
/// OAuth2 <c>client_credentials</c> token provider — direct port of the TS SDK's
/// <c>TokenProvider</c> so the .NET SDK authenticates against api.smoo.ai exactly
/// like every other SmooAI client. The token is cached in memory until 60s before
/// expiry, then refreshed; concurrent callers during a refresh share one in-flight
/// request. Consulted at request time by <see cref="Otel.AuthHeaderHandler"/> — no
/// header snapshot, no staleness.
/// </summary>
public sealed class TokenProvider
{
    private readonly string _authUrl;
    private readonly string _clientId;
    private readonly string _clientSecret;
    private readonly int _refreshWindowSec;
    private readonly HttpClient _httpClient;
    private readonly bool _ownsHttpClient;
    private readonly SemaphoreSlim _refreshGate = new(1, 1);

    private string? _accessToken;
    private long _expiresAtUnixSec;
    private Func<long> _nowUnixSec = () => DateTimeOffset.UtcNow.ToUnixTimeSeconds();

    /// <summary>
    /// Create a token provider. If <paramref name="httpClient"/> is null a private
    /// <see cref="HttpClient"/> is created and disposed with this provider.
    /// </summary>
    public TokenProvider(TokenProviderOptions options, HttpClient? httpClient = null)
    {
        ArgumentNullException.ThrowIfNull(options);
        if (string.IsNullOrEmpty(options.AuthUrl))
        {
            throw new ArgumentException("TokenProvider requires authUrl.", nameof(options));
        }
        if (string.IsNullOrEmpty(options.ClientId))
        {
            throw new ArgumentException("TokenProvider requires clientId.", nameof(options));
        }
        if (string.IsNullOrEmpty(options.ClientSecret))
        {
            throw new ArgumentException("TokenProvider requires clientSecret.", nameof(options));
        }
        _authUrl = options.AuthUrl.TrimEnd('/');
        _clientId = options.ClientId;
        _clientSecret = options.ClientSecret;
        _refreshWindowSec = options.RefreshWindowSec;
        _ownsHttpClient = httpClient is null;
        _httpClient = httpClient ?? new HttpClient();
    }

    /// <summary>
    /// Returns a valid access token, refreshing if the cached value is missing,
    /// expired, or within the refresh window. Concurrent callers during a refresh
    /// share a single token exchange.
    /// </summary>
    public async Task<string> GetAccessTokenAsync(CancellationToken cancellationToken = default)
    {
        if (!ShouldRefresh())
        {
            return _accessToken!;
        }

        await _refreshGate.WaitAsync(cancellationToken).ConfigureAwait(false);
        try
        {
            // Re-check under the lock — another caller may have refreshed.
            if (!ShouldRefresh())
            {
                return _accessToken!;
            }
            return await RefreshAsync(cancellationToken).ConfigureAwait(false);
        }
        finally
        {
            _refreshGate.Release();
        }
    }

    /// <summary>
    /// Drop the cached token. Call when an export observes a 401 so the next
    /// attempt re-mints.
    /// </summary>
    public void Invalidate()
    {
        _accessToken = null;
        _expiresAtUnixSec = 0;
    }

    private bool ShouldRefresh()
    {
        if (_accessToken is null)
        {
            return true;
        }
        return _nowUnixSec() >= _expiresAtUnixSec - _refreshWindowSec;
    }

    private async Task<string> RefreshAsync(CancellationToken cancellationToken)
    {
        using var content = new FormUrlEncodedContent(new[]
        {
            new KeyValuePair<string, string>("grant_type", "client_credentials"),
            new KeyValuePair<string, string>("provider", "client_credentials"),
            new KeyValuePair<string, string>("client_id", _clientId),
            new KeyValuePair<string, string>("client_secret", _clientSecret),
        });

        using var response = await _httpClient.PostAsync($"{_authUrl}/token", content, cancellationToken).ConfigureAwait(false);
        if (!response.IsSuccessStatusCode)
        {
            var body = await SafeReadBodyAsync(response).ConfigureAwait(false);
            throw new InvalidOperationException($"OAuth token exchange failed: HTTP {(int)response.StatusCode} {body}");
        }

        var json = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
        using var doc = JsonDocument.Parse(json);
        var root = doc.RootElement;
        if (!root.TryGetProperty("access_token", out var tokenEl) || tokenEl.ValueKind != JsonValueKind.String)
        {
            throw new InvalidOperationException("OAuth token endpoint returned no access_token.");
        }

        var token = tokenEl.GetString()!;
        var expiresIn = root.TryGetProperty("expires_in", out var expEl) && expEl.ValueKind == JsonValueKind.Number
            ? expEl.GetInt64()
            : 3600;

        _accessToken = token;
        _expiresAtUnixSec = _nowUnixSec() + expiresIn;
        return token;
    }

    private static async Task<string> SafeReadBodyAsync(HttpResponseMessage response)
    {
        try
        {
            return await response.Content.ReadAsStringAsync().ConfigureAwait(false);
        }
        catch
        {
            return "<unreadable>";
        }
    }

    /// <summary>Test seam — override the time source.</summary>
    internal void SetNowForTests(Func<long> nowUnixSec) => _nowUnixSec = nowUnixSec;

    /// <summary>Dispose the owned HttpClient, if any.</summary>
    internal void DisposeOwned()
    {
        if (_ownsHttpClient)
        {
            _httpClient.Dispose();
        }
        _refreshGate.Dispose();
    }
}

File: go/token_provider.go
package observability

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"
)

// TokenProvider is an OAuth2 client_credentials token provider — a direct port
// of the TS auth/token-provider.ts. Consulted at request time by the
// auth-injecting OTLP exporter, so there's no header snapshot / expiry drift.
// Caches in memory until refreshWindow before expiry, then refreshes;
// concurrent callers during a refresh share one in-flight request.
//
// Server contract:
//
//	POST {authUrl}/token
//	Content-Type: application/x-www-form-urlencoded
//	grant_type=client_credentials&provider=client_credentials&client_id=...&client_secret=...
type TokenProvider struct {
	authURL       string
	clientID      string
	clientSecret  string
	refreshWindow time.Duration
	client        *http.Client
	now           func() time.Time

	mu          sync.Mutex
	accessToken string
	expiresAt   time.Time
	inflight    *inflightToken
}

type inflightToken struct {
	done  chan struct{}
	token string
	err   error
}

// TokenProviderOptions configures a TokenProvider.
type TokenProviderOptions struct {
	// AuthURL is the OAuth issuer base URL, e.g. https://auth.smoo.ai.
	AuthURL string
	// ClientID is the M2M client id.
	ClientID string
	// ClientSecret is the M2M client secret (sk_...).
	ClientSecret string
	// RefreshWindow is how long before expiry to proactively refresh.
	// Defaults to 60s, matching the TS SDK.
	RefreshWindow time.Duration
	// HTTPClient overrides the default client (test seam).
	HTTPClient *http.Client
	// Now overrides the time source (test seam).
	Now func() time.Time
}

// NewTokenProvider validates options and returns a provider.
func NewTokenProvider(opts TokenProviderOptions) (*TokenProvider, error) {
	if opts.AuthURL == "" {
		return nil, errors.New("observability: TokenProvider requires AuthURL")
	}
	if opts.ClientID == "" {
		return nil, errors.New("observability: TokenProvider requires ClientID")
	}
	if opts.ClientSecret == "" {
		return nil, errors.New("observability: TokenProvider requires ClientSecret")
	}
	window := opts.RefreshWindow
	if window <= 0 {
		window = 60 * time.Second
	}
	client := opts.HTTPClient
	if client == nil {
		client = &http.Client{Timeout: 10 * time.Second}
	}
	now := opts.Now
	if now == nil {
		now = time.Now
	}
	return &TokenProvider{
		authURL:       strings.TrimRight(opts.AuthURL, "/"),
		clientID:      opts.ClientID,
		clientSecret:  opts.ClientSecret,
		refreshWindow: window,
		client:        client,
		now:           now,
	}, nil
}

// AccessToken returns a valid token, refreshing if the cached value is missing,
// expired, or within RefreshWindow of expiry. Concurrent callers during a
// refresh share one in-flight request.
func (p *TokenProvider) AccessToken(ctx context.Context) (string, error) {
	p.mu.Lock()
	if !p.shouldRefreshLocked() {
		tok := p.accessToken
		p.mu.Unlock()
		return tok, nil
	}
	if p.inflight != nil {
		fl := p.inflight
		p.mu.Unlock()
		select {
		case <-fl.done:
			return fl.token, fl.err
		case <-ctx.Done():
			return "", ctx.Err()
		}
	}
	fl := &inflightToken{done: make(chan struct{})}
	p.inflight = fl
	p.mu.Unlock()

	token, err := p.refresh(ctx)

	p.mu.Lock()
	fl.token, fl.err = token, err
	close(fl.done)
	p.inflight = nil
	p.mu.Unlock()
	return token, err
}

// Invalidate drops the cached token so the next AccessToken re-mints. Called by
// the exporter's 401 retry path.
func (p *TokenProvider) Invalidate() {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.accessToken = ""
	p.expiresAt = time.Time{}
}

func (p *TokenProvider) shouldRefreshLocked() bool {
	if p.accessToken == "" {
		return true
	}
	return !p.now().Before(p.expiresAt.Add(-p.refreshWindow))
}

func (p *TokenProvider) refresh(ctx context.Context) (string, error) {
	form := url.Values{
		"grant_type":    {"client_credentials"},
		"provider":      {"client_credentials"},
		"client_id":     {p.clientID},
		"client_secret": {p.clientSecret},
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, p.authURL+"/token", strings.NewReader(form.Encode()))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")

	resp, err := p.client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(io.LimitReader(resp.Body, 1024))
		return "", fmt.Errorf("observability: OAuth token exchange failed: HTTP %d %s", resp.StatusCode, string(body))
	}

	var parsed struct {
		AccessToken string `json:"access_token"`
		ExpiresIn   int    `json:"expires_in"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&parsed); err != nil {
		return "", fmt.Errorf("observability: OAuth token response decode failed: %w", err)
	}
	if parsed.AccessToken == "" {
		return "", errors.New("observability: OAuth token endpoint returned no access_token")
	}
	expiresIn := parsed.ExpiresIn
	if expiresIn <= 0 {
		expiresIn = 3600
	}

	p.mu.Lock()
	p.accessToken = parsed.AccessToken
	p.expiresAt = p.now().Add(time.Duration(expiresIn) * time.Second)
	p.mu.Unlock()
	return parsed.AccessToken, nil
}
